File: modules/student-experience/src/main/java/com/campusos/portal/payload/PayloadCodec.java
package com.campusos.portal.payload;

import com.campusos.portal.domain.RequestType;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;

/** Typed DTO in, JSON string out. Nothing writes `payload` without passing through here. */
@Component
public class PayloadCodec {

    private final ObjectMapper mapper = new ObjectMapper();

    public String write(RequestPayload p) {
        p.validate();
        try {
            return mapper.writeValueAsString(p);
        } catch (Exception e) {
            throw new IllegalStateException("payload serialize failed", e);
        }
    }

    public RequestPayload read(RequestType type, String json) {
        try {
            return switch (type) {
                case LEAVE -> mapper.readValue(json, LeavePayload.class);
                case INTERNSHIP -> mapper.readValue(json, InternshipPayload.class);
                case DOCUMENT -> mapper.readValue(json, DocumentPayload.class);
                case GRIEVANCE -> mapper.readValue(json, GrievancePayload.class);
            };
        } catch (Exception e) {
            throw new IllegalStateException("payload deserialize failed for " + type, e);
        }
    }
}

File: modules/student-experience/src/payload/PayloadCodec.js
import { RequestType } from '../domain/enums.js';
import { LeavePayload } from './LeavePayload.js';
import { InternshipPayload } from './InternshipPayload.js';
import { DocumentPayload } from './DocumentPayload.js';
import { GrievancePayload } from './GrievancePayload.js';

/** Typed DTO in, JSON string out. Nothing writes `payload` without passing through here. */
export const PayloadCodec = {
  write(p) {
    p.validate();
    return JSON.stringify(p);
  },

  read(type, json) {
    const o = JSON.parse(json);
    switch (type) {
      case RequestType.LEAVE: return LeavePayload.fromJSON(o);
      case RequestType.INTERNSHIP: return InternshipPayload.fromJSON(o);
      case RequestType.DOCUMENT: return DocumentPayload.fromJSON(o);
      case RequestType.GRIEVANCE: return GrievancePayload.fromJSON(o);
      default: throw new Error(`payload deserialize failed for ${type}`);
    }
  },
};
